pkg/controller/job/job_controller.go

```diff
@@ -0,0 +1,297 @@
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	job, err := v.jobLister.Jobs(namespace).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return err
+		}
+		log.Error(err, "get job failed", "namespace", namespace, "name", name)
+		return nil
+	}
+
+	err = v.makeRevision(job)
+
+	if err != nil {
+		log.Error(err, "make job revision failed", "namespace", namespace, "name", name)
+		return nil
+	}
+
+	return nil
+}
+
+// When a job is added, figure out which service it will be used
+// and enqueue it. obj must have *batchv1.Job type
+func (v *JobController) addJob(obj interface{}) {
+	deploy := obj.(*batchv1.Job)
+
+	v.queue.Add(deploy.Name)
+
+	return
+}
+
+func (v *JobController) handleErr(err error, key interface{}) {
+	if err != nil {
+		v.queue.Forget(key)
+		return
+	}
+
+	if v.queue.NumRequeues(key) < maxRetries {
+		log.V(2).Info("Error syncing virtualservice for service, retrying.", "key", key, "error", err)
+		v.queue.AddRateLimited(key)
+		return
+	}
+
+	log.V(4).Info("Dropping service out of the queue", "key", key, "error", err)
+	v.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (v *JobController) makeRevision(job *batchv1.Job) error {
+	revisionIndex := -1
+	revisions, err := v.getRevisions(job)
+
+	// failed get revisions
+	if err != nil {
+		return nil
+	}
+
+	uid := job.UID
+	for index, revision := range revisions {
+		if revision.Uid == string(uid) {
+			currentRevision := v.getCurrentRevision(job)
+			if reflect.DeepEqual(currentRevision, revision) {
+				return nil
+			} else {
+				revisionIndex = index
+				break
+			}
+		}
+	}
+
+	if revisionIndex == -1 {
+		revisionIndex = len(revisions) + 1
+	}
+
+	revisions[revisionIndex] = v.getCurrentRevision(job)
+
+	revisionsByte, err := json.Marshal(revisions)
+	if err != nil {
+		glog.Error(err)
+	}
+
+	if job.Annotations == nil {
+		job.Annotations = make(map[string]string)
+	}
+
+	job.Annotations[revisionsAnnotationKey] = string(revisionsByte)
+	_, err = v.client.BatchV1().Jobs(job.Namespace).Update(job)
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (v *JobController) getRevisions(job *batchv1.Job) (JobRevisions, error) {
+	revisions := make(JobRevisions)
+
+	if revisionsStr := job.Annotations["revisions"]; revisionsStr != "" {
+		err := json.Unmarshal([]byte(revisionsStr), &revisions)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get job %s's revisions, reason: %s", job.Name, err)
+		}
+	}
+
+	return revisions, nil
+}
+
+func (v *JobController) getCurrentRevision(item *batchv1.Job) JobRevision {
+	var revision JobRevision
+	for _, condition := range item.Status.Conditions {
+		if condition.Type == "Failed" && condition.Status == "True" {
+			revision.Status = Failed
+			revision.Reasons = append(revision.Reasons, condition.Reason)
+			revision.Messages = append(revision.Messages, condition.Message)
+		}
+
+		if condition.Type == "Complete" && condition.Status == "True" {
```

Comment:
is that possible item.Status.Conditions contains both Failed condtion.Type and Complete condition.Type? If true, here will be a problem, otherwise, use a elif to join these two `if` blocks to make it more clear